sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/UnboundedReaderImpl.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.UnboundedSource.UnboundedReader;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+
+public class UnboundedReaderImpl extends UnboundedReader<SequencedMessage> {
+
+  private final UnboundedSource<SequencedMessage, CheckpointMarkImpl> source;
+  private final MemoryBufferedSubscriber subscriber;
+  private final TopicBacklogReader backlogReader;
+  private final BlockingCommitter committer;
+
+  private Offset fetchOffset;
+  private Optional<Instant> lastMessageTimestamp = Optional.empty();
+  private boolean advanced = false;
+
+  UnboundedReaderImpl(
+      UnboundedSource<SequencedMessage, CheckpointMarkImpl> source,
+      MemoryBufferedSubscriber subscriber,
+      TopicBacklogReader backlogReader,
+      BlockingCommitter committer,
+      Offset initialOffset) {
+    checkArgument(initialOffset.equals(subscriber.fetchOffset()));
+    this.source = source;
+    this.subscriber = subscriber;
+    this.backlogReader = backlogReader;
+    this.committer = committer;
+    this.fetchOffset = initialOffset;
+  }
+
+  @Override
+  public SequencedMessage getCurrent() throws NoSuchElementException {
+    if (!advanced) {
+      throw new NoSuchElementException();
+    }
+    return subscriber.peek().get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() throws NoSuchElementException {
+    return getTimestamp(getCurrent());
+  }
+
+  private static Instant getTimestamp(SequencedMessage message) {
+    return Instant.ofEpochMilli(Timestamps.toMillis(message.getPublishTime()));
+  }
+
+  @Override
+  public void close() throws IOException {
+    try (AutoCloseable c1 = backlogReader;
+        AutoCloseable c2 = committer;
+        AutoCloseable c3 = asCloseable(subscriber)) {
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      subscriber.startAsync().awaitRunning(1, TimeUnit.MINUTES);
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    if (!subscriber.state().equals(State.RUNNING)) {
+      throw new IOException("Subscriber failed: ", subscriber.failureCause());
+    }
+    if (advanced) {
+      subscriber.pop();
+    }
+    Optional<SequencedMessage> next = subscriber.peek();
+    advanced = next.isPresent();
+    if (!advanced) {
+      return false;
+    }
+    Offset nextOffset = Offset.of(next.get().getCursor().getOffset() + 1);
+    checkState(nextOffset.value() > fetchOffset.value());
+    fetchOffset = nextOffset;
+    lastMessageTimestamp = Optional.of(getTimestamp(next.get()));
+    return true;
+  }
+
+  @Override
+  public Instant getWatermark() {
+    return lastMessageTimestamp.orElse(BoundedWindow.TIMESTAMP_MIN_VALUE);
```

Comment:
are timestamps increasing in PubsubLite?  Or does this need to be based upon last messages and some controllable skew option etc

Reply:
Timestamps are always increasing or equal, we reference truetime for this